Fuzzysearch

philipanda opened the pull request.

philipanda committed:
utils/fuzzy.py: Add

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

philipanda committed:
libraries/Telnet.py: Add fuzzy string comparisons

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

philipanda committed:
libraries/String.py: Add fuzzy string comparisons

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

philipanda committed:
utils/asserts.py: Add fuzzy comparisons

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

philipanda committed:
libraries/Builtin.py: Add fuzzy comparisons

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

philipanda committed:
utest/requirements.txt: Add fuzzysearch requirement

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

philipanda committed:
Collections.py: Add get_index_from_list_fuzzy

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

philipanda committed:
Telnet.py: Export Read Until Fuzzy keyword

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

philipanda committed:
Telnet.py: Fix missing args to terminal_reauntil_fuzzy

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

philipanda committed:
Collections.py: Add List Should Contain Fuzzy

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

macpijan commented on src/robot/libraries/BuiltIn.py:
```diff
@@ -704,6 +709,31 @@ def _should_be_equal(self, first, second, msg, values, formatter="str"):
             self._raise_multi_diff(first, second, msg, formatter)
         assert_equal(first, second, msg, include_values, formatter)
+    
+    def should_be_equal_fuzzy(self, first, second, msg=None, values=True,
```
I wonder, how much resources do we need from the original classes?
If not much - maybe we want to create a separate `fuzzy` library for this purpose?

If we do - maybe it makes sense to  reuse some code and add these new `match/error` parameters as optional ones into existing keywords?

philipanda replied:
From these ones, pretty much none. Now I think it would be better to move such keywords into a separate external library to keep the robotframework fork clean in case of a need for rebase.
I think that would be a good improvement, but not strictly required for now

philipanda replied:
https://github.com/Dasharo/open-source-firmware-validation/issues/1044

philipanda committed:
Telnet.py: Add keyword decorator to new keyword

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

philipanda committed:
Telnet.py: Add _get_keywords debug prints

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

philipanda committed:
utils/fuzzy.py: Add max_insertions and max_deletions

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

philipanda committed:
libraries/Telnet.py: Add max_insertions and max_deletions to fuzzy methods

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

philipanda committed:
libraries/Strings.py: Add max_insertions and max_deletions to fuzzy methods

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

philipanda committed:
util/asserts.py: Add max_insertions and max_deletions to fuzzy methods

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

philipanda committed:
BuiltIn.py: Add max_insertions and max_deletions to fuzzy methods

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

philipanda committed:
String.py: Add max_insertions and max_deletions to fuzzy methods

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

philipanda committed:
utils/fuzzy.py: Add exception logging

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

philipanda committed:
Telnet.py: Remove debug prints

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

philipanda committed:
fuzzy.py: Add arguments debug prints

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

philipanda committed:
fuzzy.py: Add robust type checks

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

philipanda committed:
fuzzy.py: print logs only on exceptions

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

filipleple commented:
@philipanda do we know why CI fails with 

```
[ ERROR ] Importing library 'robot.libraries.BuiltIn' failed: ModuleNotFoundError: No module named 'fuzzysearch'
```

Maybe the added library is not listed/included somewhere?

philipanda committed:
atest/requirements.txt: Add fuzzysearch library

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>

philipanda commented:
@filipleple 
> @philipanda do we know why CI fails with
> 
> ```
> [ ERROR ] Importing library 'robot.libraries.BuiltIn' failed: ModuleNotFoundError: No module named 'fuzzysearch'
> ```
> 
> Maybe the added library is not listed/included somewhere?

Yup, looks like unit tests and acceptance tests use separate `requirements.txt` files. It should be good now

philipanda commented:
@filipleple looks like the problems with keeping all the python requirements in check is not unique to OSFV. 
The package was installed here:
https://github.com/3mdeb/robotframework/actions/runs/17294681794/job/49089883820?pr=22#step:10:67
But it still is not being detected in the end:
https://github.com/3mdeb/robotframework/actions/runs/17294681794/job/49089883820?pr=22#step:10:100

philipanda committed:
fuzzysearch: change max_errors to max_substitutions

setting max_errors, max_insertions and max_deletions, but
not max_substitutions allowed for an edge case to trigger
where the fuzzysearch library was crashing.

Changing the interface like that should prevent unhandled
situations

Signed-off-by: Filip Gołaś <filip.golas@3mdeb.com>
